pkg/connectorrunner/runner.go

```diff
@@ -67,51 +76,94 @@ func (c *connectorRunner) handleContextCancel(ctx context.Context) error {
 		return nil
 	}
 
-	l.Debug("unexpected context cancellation", zap.Error(err))
+	l.Debug("runner: unexpected context cancellation", zap.Error(err))
 	return err
 }
+func (c *connectorRunner) processTask(ctx context.Context, task *v1.Task) error {
+	cc, err := c.cw.C(ctx)
+	if err != nil {
+		return fmt.Errorf("runner: error creating connector client: %w", err)
+	}
+
+	err = c.tasks.Process(ctx, task, cc)
+	if err != nil {
+		return fmt.Errorf("runner: error processing task: %w", err)
+	}
+
+	return nil
+}
+
+func (c *connectorRunner) backoff(ctx context.Context, errCount int) time.Duration {
+	waitDuration := time.Duration(errCount*errCount) * time.Second
+	if waitDuration > time.Minute {
+		waitDuration = time.Minute
+	}
+	return waitDuration
+}
 
 func (c *connectorRunner) run(ctx context.Context) error {
 	l := ctxzap.Extract(ctx)
 
-	var waitDuration time.Duration
-	var nextTask *v1.Task
+	sem := semaphore.NewWeighted(int64(taskConcurrency))
+
+	waitDuration := time.Second * 0
+	errCount := 0
 	var err error
 	for {
 		select {
 		case <-ctx.Done():
 			return c.handleContextCancel(ctx)
 		case <-time.After(waitDuration):
-		}
+			l.Debug("runner: claiming worker")
```

Comment:
I'm thinking we should toss this into a function so we can `defer` the semaphore release and save us some pain at some point.